crypto/ml_kem/ml_kem.c

```diff
@@ -0,0 +1,1567 @@
+    if (!hash_g(Kr, decrypted, sizeof(decrypted), ctx)
+        || !encrypt_cpa(tmp_ctext, decrypted, m, t, r, tmp_y, tmp_e1, tmp_u,
+                        vinfo, ctx)) {
+        memcpy(shared_secret, failure_key, ML_KEM_SHARED_SECRET_BYTES);
+        return 1;
+    }
+    mask = constant_time_eq_int_8(0,
+        CRYPTO_memcmp(ctext, tmp_ctext, vinfo->ctext_bytes));
+    for (i = 0; i < ML_KEM_SHARED_SECRET_BYTES; i++)
+        shared_secret[i] = constant_time_select_8(mask, Kr[i], failure_key[i]);
+    return 1;
+}
+
+/*
+ * -----
+ *
+ * Perform ML_KEM ops for a specific parameter set, the public and private key
+ * pointers are (void *), in that order, when either is accepted in lieu of a
+ * public key of the appropriate type.  Parameters with an implicit fixed
+ * length in each variant have an explicit checked additional length argument
+ * here.
+ */
+# define ml_kem_bname(bits, b)  ossl_ml_kem_##bits##_##b
+# define encode_pubkey(bits)    ml_kem_bname(bits, encode_public_key)
+# define encode_prvkey(bits)    ml_kem_bname(bits, encode_private_key)
+# define parse_pubkey(bits)     ml_kem_bname(bits, parse_public_key)
+# define parse_prvkey(bits)     ml_kem_bname(bits, parse_private_key)
+# define genkey_rand(bits)      ml_kem_bname(bits, genkey_rand)
+# define genkey_seed(bits)      ml_kem_bname(bits, genkey_seed)
+# define encap_rand(bits)       ml_kem_bname(bits, encap_rand)
+# define encap_seed(bits)       ml_kem_bname(bits, encap_seed)
+# define decap(bits)            ml_kem_bname(bits, decap)
+# define pub_t(bits)            ml_kem_bname(bits, public_key)
+# define prv_t(bits)            ml_kem_bname(bits, private_key)
+
+/*
+ * Convert either a public or private key (void *) pointer to
+ * a pointer of the correct type with a possible "const" qualifier.
+ * In the case of public keys, we may be willing to also accept a private key,
+ * which embeds an associated public key (|pubish| and |constish|).
+ */
+# define pubcast(bits, p)       ((pub_t(bits) *) (p))
+# define pubconst(bits, p)      ((const pub_t(bits) *) (p))
+# define prvcast(bits, p)       ((prv_t(bits) *) (p))
+# define prvconst(bits, p)      ((const prv_t(bits) *) (p))
+# define pubish(bits, p, q)     pubcast((p) != NULL ? \
+                                        pubcast(bits, p) : \
+                                        &(prvcast(bits, q)->pub))
+# define constish(bits, p, q)   pubconst(bits, (p) != NULL ? \
+                                         pubconst(bits, p) : \
+                                         &(prvconst(bits, q)->pub))
+
+int ossl_ml_kem_vencode_public_key(vinfo_t v,
+                                   uint8_t *out,
+                                   size_t len,
+                                   const void *pub,
+                                   const void *prv)
+{
+# define case_encode_pubkey(bits) \
+    case ML_KEM_##bits##_RANK: \
+        if ((pub == NULL && prv == NULL) || len != v->pubkey_bytes) \
```

Comment:
The if could be done once outside of the macro..

Reply:
Same comment applies to most of these v functions.

Reply:
Fixed in #26172